File: src/data/4o/Shortest_Distance_After_Road_Addition_Queries_II/Shortest_Distance_After_Road_Addition_Queries_II.java
import java.util.*;

class Solution {
    public int[] shortestDistanceAfterQueries(int n, int[][] queries) {
        int[] dist = new int[n]; // Array to store the shortest distance from city 0 to city n-1
        Arrays.fill(dist, Integer.MAX_VALUE);
        dist[0] = 0; // Distance from city 0 to itself is 0

        List<List<int[]>> adjList = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            adjList.add(new ArrayList<>());
        }

        // Initially, there is a road from city i to city i+1
        for (int i = 0; i < n - 1; i++) {
            adjList.get(i).add(new int[]{i + 1, 1}); // Distance of each initial road is 1
        }

        // Result array to store the answer after each query
        int[] result = new int[queries.length];

        // Process each query
        for (int q = 0; q < queries.length; q++) {
            int u = queries[q][0];
            int v = queries[q][1];
            adjList.get(u).add(new int[]{v, 1}); // Add the new road with distance 1

            // Recompute the shortest path from city 0 to city n-1 using Dijkstra's algorithm
            PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[1] - b[1]); // (city, distance)
            Arrays.fill(dist, Integer.MAX_VALUE);
            dist[0] = 0;
            pq.add(new int[]{0, 0});

            while (!pq.isEmpty()) {
                int[] current = pq.poll();
                int city = current[0];
                int distance = current[1];

                if (distance > dist[city]) continue;

                for (int[] neighbor : adjList.get(city)) {
                    int nextCity = neighbor[0];
                    int newDist = distance + neighbor[1];

                    if (newDist < dist[nextCity]) {
                        dist[nextCity] = newDist;
                        pq.add(new int[]{nextCity, newDist});
                    }
                }
            }

            result[q] = dist[n - 1] == Integer.MAX_VALUE ? -1 : dist[n - 1];
        }

        return result;
    }
}

public class Shortest_Distance_After_Road_Addition_Queries_II {
public static void runTests(Solution solution) {
    // Test case 1: Simple addition that extends path
    int n1 = 5;
    int[][] queries1 = {{2, 4}, {0, 2}, {0, 4}};
    int[] expected_result1 = {3, 2, 1};
    assert java.util.Arrays.equals(solution.shortestDistanceAfterQueries(n1, queries1), expected_result1) : "Test case 1 failed. Expected " + java.util.Arrays.toString(expected_result1);

    // Test case 2: Additions that skip intermediate cities
    int n2 = 4;
    int[][] queries2 = {{0, 3}, {0, 2}};
    int[] expected_result2 = {1, 1};
    assert java.util.Arrays.equals(solution.shortestDistanceAfterQueries(n2, queries2), expected_result2) : "Test case 2 failed. Expected " + java.util.Arrays.toString(expected_result2);

    // Test case 3: No additional road added
    int n3 = 3;
    int[][] queries3 = {};
    int[] expected_result3 = {};
    assert java.util.Arrays.equals(solution.shortestDistanceAfterQueries(n3, queries3), expected_result3) : "Test case 3 failed. Expected " + java.util.Arrays.toString(expected_result3);

    // Test case 4: Large n with direct connection from start to end
    int n4 = 6;
    int[][] queries4 = {{0, 5}};
    int[] expected_result4 = {1};
    assert java.util.Arrays.equals(solution.shortestDistanceAfterQueries(n4, queries4), expected_result4) : "Test case 4 failed. Expected " + java.util.Arrays.toString(expected_result4);

    // Test case 5: Multiple roads but none affecting shortest path
    int n5 = 5;
    int[][] queries5 = {{0, 2}, {1, 3}, {2, 4}};
    int[] expected_result5 = {3, 3, 2};
    assert java.util.Arrays.equals(solution.shortestDistanceAfterQueries(n5, queries5), expected_result5) : "Test case 5 failed. Expected " + java.util.Arrays.toString(expected_result5);

    // Test case 6: Backward connections with no effect
    int n6 = 4;
    int[][] queries6 = {{2, 0}, {3, 1}};
    int[] expected_result6 = {3, 3};
    assert java.util.Arrays.equals(solution.shortestDistanceAfterQueries(n6, queries6), expected_result6) : "Test case 6 failed. Expected " + java.util.Arrays.toString(expected_result6);

    // Test case 7: Multiple additions creating overlapping paths
    int n7 = 7;
    int[][] queries7 = {{0, 3}, {2, 5}, {4, 6}};
    int[] expected_result7 = {4, 4, 3};
    assert java.util.Arrays.equals(solution.shortestDistanceAfterQueries(n7, queries7), expected_result7) : "Test case 7 failed. Expected " + java.util.Arrays.toString(expected_result7);
}


    public static void main(String[] args) {
        Solution solution = new Solution();
        for (int i = 0; i < 10000000; i++) {
            runTests(solution); // Ensure your test cases are implemented inside runTests
        }
    }
}

    

File: src/data/4o/Shortest_Distance_After_Road_Addition_Queries_II/Shortest_Distance_After_Road_Addition_Queries_II.cpp
#include <vector>
#include <queue>
#include <climits>
using namespace std;

class Solution {
public:
    vector<int> shortestDistanceAfterQueries(int n, vector<vector<int>>& queries) {
        vector<int> answer;
        vector<int> dist(n, INT_MAX);
        dist[0] = 0;
        vector<vector<int>> adj(n);

        // Initially, add the direct roads from city i to city i + 1
        for (int i = 0; i < n - 1; ++i) {
            adj[i].push_back(i + 1);
        }

        // Dijkstra's algorithm to update distances with priority queue
        auto dijkstra = [&]() {
            vector<int> newDist(n, INT_MAX);
            newDist[0] = 0;
            priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
            pq.push({0, 0});

            while (!pq.empty()) {
                int d = pq.top().first;
                int u = pq.top().second;
                pq.pop();

                if (d > newDist[u]) continue;

                for (int v : adj[u]) {
                    int cost = 1;
                    if (newDist[u] + cost < newDist[v]) {
                        newDist[v] = newDist[u] + cost;
                        pq.push({newDist[v], v});
                    }
                }
            }

            return newDist[n - 1];
        };

        for (auto& query : queries) {
            int u = query[0], v = query[1];
            adj[u].push_back(v);
            answer.push_back(dijkstra());
        }

        return answer;
    }
};


//  Test cases for the C++ solution
#include <cassert>

// Test cases for the C++ solution
void testSolution(Solution& solution) {
    // Test case 1: Simple addition that extends path
    int n = 5;
    std::vector<std::vector<int>> queries = {{2, 4}, {0, 2}, {0, 4}};
    std::vector<int> expected_result = {3, 2, 1};  // Path: 0 -> 1 -> 2 -> 3 -> 4, adding shortcuts but shortest path remains 4
    assert(solution.shortestDistanceAfterQueries(n, queries) == expected_result && "Test case 1 failed.");

    // Test case 2: Additions that skip intermediate cities
    n = 4;
    queries = {{0, 3}, {0, 2}};
    expected_result = {1, 1};  // Path: 0 -> 2 -> 3; shortest path remains at length 3
    assert(solution.shortestDistanceAfterQueries(n, queries) == expected_result && "Test case 2 failed.");

    // Test case 3: No additional road added
    n = 3;
    queries = {};
    expected_result = {};  // No queries to process
    assert(solution.shortestDistanceAfterQueries(n, queries) == expected_result && "Test case 3 failed.");

    // Test case 4: Large n with direct connection from start to end
    n = 6;
    queries = {{0, 5}};
    expected_result = {1};  // Path: 0 -> 5 directly; shortest path is 1
    assert(solution.shortestDistanceAfterQueries(n, queries) == expected_result && "Test case 4 failed.");

    // Test case 5: Multiple roads but none affecting shortest path
    n = 5;
    queries = {{0, 2}, {1, 3}, {2, 4}};
    expected_result = {3, 3, 2};  // Original path 0 -> 1 -> 2 -> 3 -> 4; shortcuts do not reduce length
    assert(solution.shortestDistanceAfterQueries(n, queries) == expected_result && "Test case 5 failed.");

    // Test case 6: Backward connections with no effect
    n = 4;
    queries = {{2, 0}, {3, 1}};
    expected_result = {3, 3};  // Adding backward edges, shortest path remains same 0 -> 1 -> 2 -> 3
    assert(solution.shortestDistanceAfterQueries(n, queries) == expected_result && "Test case 6 failed.");

    // Test case 7: Multiple additions creating overlapping paths
    n = 7;
    queries = {{0, 3}, {2, 5}, {4, 6}};
    expected_result = {4, 4, 3};  // Path: 0 -> 3 -> 5 -> 6; remains unaffected by direct connections
    assert(solution.shortestDistanceAfterQueries(n, queries) == expected_result && "Test case 7 failed.");
}

int main() {
    Solution solution;
    for (int i = 0; i < 327680; ++i) {
        testSolution(solution);  
    }
    return 0;
}
